### PR TITLE
NEXUS-7933: Implement the "fallback" that Javadoc mentions

### PR DESCRIPTION
**Pull request by cstamas**
In case of non-empty directory cross-volume move, the code was
incorrect, and did not fulfill contract described in method
javadoc, there was no "fallback".

Issue
https://issues.sonatype.org/browse/NEXUS-7933

CI
http://bamboo.s/browse/NX-OSSF572

### CODE REVIEW

**Comment by mcculls** [comment]
+1 looks reasonable, hopefully CI gets well soon ;)

File: components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java [hunk 1]
```diff
@@ -402,7 +434,11 @@ public static boolean copyIfExists(final Path from, final Path to, final @Nullab
   public static void move(final Path from, final Path to)
       throws IOException
   {
-    Files.move(from, to, StandardCopyOption.REPLACE_EXISTING);
+    if (areOnSameFileStore(from, to)) {
+      sameFileStoreMove(from, to);
```
Comment (mcculls) [inline components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java]: Maybe consider falling back to `crossFileStoreMove` if this throws an `IOException`?  I believe the code used to do that before it was refactored, but I might be mis-remembering.

File: components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java [hunk 2]
```diff
@@ -394,6 +399,45 @@ public static boolean copyIfExists(final Path from, final Path to, final @Nullab
...
+      final FileStore fromStore = Files.getFileStore(from); // from must exist
+      Path toExistingParent = to.normalize(); // to usually does not exists, is about to be created as part of move
+      while (toExistingParent != null && !Files.exists(toExistingParent)) {
+        toExistingParent = toExistingParent.getParent();
+      }
+      if (toExistingParent != null) {
+        final FileStore toStore = Files.getFileStore(toExistingParent);
+        return fromStore.equals(toStore);
+      } else {
+        log.warn("No parent path found for target while move '{}'->'{}'", from, to, new RuntimeException("marker")); // record the stack trace?
```
Comment (peterlynch) [inline components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java]: minor: "during move" instead of "while move"

**Comment by kellyrob99** [comment]
+1

File: components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java [hunk 3]
```diff
@@ -394,6 +399,45 @@ public static boolean copyIfExists(final Path from, final Path to, final @Nullab
...
+  private static boolean areOnSameFileStore(final Path from, final Path to) {
+    try {
+      final FileStore fromStore = Files.getFileStore(from); // from must exist
+      Path toExistingParent = to.normalize(); // to usually does not exists, is about to be created as part of move
+      while (toExistingParent != null && !Files.exists(toExistingParent)) {
+        toExistingParent = toExistingParent.getParent();
+      }
+      if (toExistingParent != null) {
+        final FileStore toStore = Files.getFileStore(toExistingParent);
+        return fromStore.equals(toStore);
```
Comment (jdillon) [inline components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java]: Is it documented anywhere that this is how you detect different underlying filesystems?

File: components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java (hunk 2 above)
Comment (jdillon) [inline components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java]: method doesn't imply move, so warning may be better of generic

File: components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java (hunk 3 above)
Comment (mcculls) [inline components/nexus-core/src/main/java/org/sonatype/nexus/util/file/DirSupport.java]: Couldn't find definitive documentation, but while searching just now I did notice a similar equality test in logback:

https://github.com/qos-ch/logback/blob/master/logback-core/src/main/java/ch/qos/logback/core/rolling/helper/RenameUtil.java#L67 ->
https://github.com/qos-ch/logback/blob/master/logback-core/src/main/java/ch/qos/logback/core/rolling/helper/RenameUtil.java#L97 ->
https://github.com/qos-ch/logback/blob/master/logback-core/src/main/java/ch/qos/logback/core/rolling/helper/FileStoreUtil.java#L70

**Comment by jdillon** [comment]
+1 if it works in our testing.  Still not really fond of the FileStore check, but it may be as expected.  But if it works, then lets move forward.

**Comment by cstamas** [comment]
Manually merged as de56411807c54d1ed154ff2d4f13274a41e65af3

**Pull request closed without merging**